File: solutions/2026-09-10/cpp/2026-09-10-count-nodes-equal-to-average-of-subtree.cpp
```cpp
/**
 * @brief Counts the number of nodes in a binary tree whose value equals the average of their subtree.
 * 
 * @link https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/
 * 
 * @approach
 * We can use a Depth-First Search (DFS) approach to solve this problem.
 * For each node, we need to calculate the sum of all values in its subtree and the count of nodes in its subtree.
 * A post-order traversal is suitable for this. We define a recursive helper function that returns a pair:
 * {sum of subtree, count of subtree nodes}.
 * Inside the helper function:
 * 1. Recursively call the function for the left child and store its results.
 * 2. Recursively call the function for the right child and store its results.
 * 3. Calculate the current node's subtree sum by adding its own value to the sums returned from its children.
 * 4. Calculate the current node's subtree node count by adding 1 (for the current node) to the counts returned from its children.
 * 5. Calculate the average of the current subtree: sum / count.
 * 6. If the node's value equals this average, increment a global counter.
 * 7. Return the calculated subtree sum and count for the current node.
 * 
 * The base case for the recursion is a null node, which returns {0, 0}.
 * 
 * @time_complexity O(N), where N is the number of nodes in the tree. Each node is visited exactly once during the DFS traversal.
 * @space_complexity O(H), where H is the height of the tree. This is due to the recursion stack depth. In the worst case (a skewed tree), H can be N. In the best case (a balanced tree), H is log N.
 */

#include <iostream>
#include <vector>
#include <numeric>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int count = 0; // Global counter to store the number of nodes that satisfy the condition.

    // Helper function for DFS traversal.
    // Returns a pair: {sum of subtree, count of nodes in subtree}.
    std::pair<int, int> dfs(TreeNode* node) {
        // Base case: if the node is null, return {0, 0} (no sum, no nodes).
        if (!node) {
            return {0, 0};
        }

        // Recursively call DFS for the left child.
        std::pair<int, int> left_subtree_info = dfs(node->left);
        // Recursively call DFS for the right child.
        std::pair<int, int> right_subtree_info = dfs(node->right);

        // Calculate the sum of the current subtree.
        // It's the current node's value + sum from the left subtree + sum from the right subtree.
        int current_subtree_sum = node->val + left_subtree_info.first + right_subtree_info.first;
        
        // Calculate the count of nodes in the current subtree.
        // It's 1 (for the current node) + count from the left subtree + count from the right subtree.
        int current_subtree_count = 1 + left_subtree_info.second + right_subtree_info.second;

        // Calculate the average of the current subtree.
        // The problem states to round down to the nearest integer, which integer division naturally does.
        int average = current_subtree_sum / current_subtree_count;

        // Check if the current node's value is equal to the calculated average.
        if (node->val == average) {
            count++; // If it is, increment the global counter.
        }

        // Return the sum and count of the current subtree for its parent to use.
        return {current_subtree_sum, current_subtree_count};
    }

    // Main function to count nodes equal to the average of their subtree.
    int averageOfSubtree(TreeNode* root) {
        count = 0; // Reset the counter for each new call.
        dfs(root); // Start the DFS traversal from the root.
        return count; // Return the final count.
    }
};
```

File: solutions/2026-09-10/javascript/2026-09-10-count-nodes-equal-to-average-of-subtree.js
/**
 * @file LeetCode Problem: Count Nodes Equal to Average of Subtree
 * @author Your Name
 * @link https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/
 * @summary Counts nodes whose value equals the average of their subtree.
 *
 * Approach:
 * We can use a Depth-First Search (DFS) approach to traverse the tree. For each node,
 * we need to calculate the sum of values and the count of nodes in its subtree.
 * A recursive DFS function can return a pair: [sum of subtree, count of subtree].
 * Inside the DFS function, we recursively call it for the left and right children.
 * The current node's sum will be its value plus the sums from its children.
 * The current node's count will be 1 plus the counts from its children.
 * After calculating the subtree sum and count for a node, we check if the node's value
 * is equal to the floor of (subtree sum / subtree count). If it is, we increment a global counter.
 *
 * Time Complexity:
 * O(N), where N is the number of nodes in the tree. We visit each node exactly once
 * during the DFS traversal. For each node, we perform constant time operations (addition, division, comparison).
 *
 * Space Complexity:
 * O(H), where H is the height of the tree. This is due to the recursion stack depth.
 * In the worst case (a skewed tree), H can be N, resulting in O(N) space.
 * In the best case (a balanced tree), H is logN, resulting in O(logN) space.
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

// Global variable to store the count of nodes that satisfy the condition.
let count = 0;

/**
 * Counts nodes where the node's value is equal to the average of its subtree.
 * @param {TreeNode} root The root of the binary tree.
 * @return {number} The number of nodes satisfying the condition.
 */
var averageOfSubtree = function(root) {
    // Reset the global count for each new test case.
    count = 0;
    // Start the DFS traversal from the root.
    dfs(root);
    // Return the final count.
    return count;
};

/**
 * Performs a Depth-First Search traversal to calculate subtree sums, counts,
 * and check the average condition for each node.
 * @param {TreeNode} node The current node being visited.
 * @return {[number, number]} A pair containing [sum of subtree, count of subtree].
 */
function dfs(node) {
    // Base case: If the node is null, it contributes 0 sum and 0 count.
    if (!node) {
        return [0, 0];
    }

    // Recursively call DFS for the left and right children.
    const [leftSum, leftCount] = dfs(node.left);
    const [rightSum, rightCount] = dfs(node.right);

    // Calculate the total sum of the current node's subtree.
    // It's the current node's value plus the sums from its left and right subtrees.
    const currentSum = node.val + leftSum + rightSum;

    // Calculate the total count of nodes in the current node's subtree.
    // It's 1 (for the current node) plus the counts from its left and right subtrees.
    const currentCount = 1 + leftCount + rightCount;

    // Calculate the average of the subtree, rounded down.
    const average = Math.floor(currentSum / currentCount);

    // Check if the current node's value is equal to the calculated average.
    if (node.val === average) {
        // If it is, increment the global count.
        count++;
    }

    // Return the sum and count of the current subtree to its parent.
    return [currentSum, currentCount];
}
